randovania/generator/filler/retcon.py

```diff
@@ -85,6 +85,41 @@ def _resources_in_pickup(pickup: PickupEntry, current_resources: CurrentResource
 def _should_have_hint(item_category: ItemCategory) -> bool:
     return item_category.is_major_category or item_category == ItemCategory.TEMPLE_KEY
 
+def _place_hints(final_state: State, rng: Random, status_update: Callable[[str], None]) -> GamePatches:
+    status_update("Placing hints...")
+
+    patches = final_state.patches
+
+    state_sequence = [final_state]
+    while state_sequence[-1].previous_state:
+        state_sequence.append(state_sequence[-1].previous_state)
+    state_sequence = tuple(reversed(state_sequence))
+
+    indices_with_hints = {index for index in final_state.patches.pickup_assignment if
+                          _should_have_hint(final_state.patches.pickup_assignment[index].item_category)}
+
+    sequence = []
+    for state in state_sequence[1:]:
+        new_indices = list((set(state.collected_pickup_indices) & indices_with_hints)
+                           - (set(state.previous_state.collected_pickup_indices) | set(patches.starting_items)))
```

Comment:
`patches.starting_items` should only contains resources of the type `ITEM`. I don't think you need to use it here.